add persons api

SonQBChau opened the pull request:
## What's new
- add person apis
- add db_test.sql script to insert sample data into the database from sql

### How to test
- set .env LOAD_VOCAB=true
- rebuild the stack
- run db_test.sql inside postgres

SonQBChau committed:
add persons api

- add new endpoint to schema to create new person only
- add all the endpoints for person

SonQBChau committed:
add person_id auto-increment

SonQBChau committed:
Update db_setup.py

SonQBChau committed:
Update db_setup.py

SonQBChau committed:
Update app.py

SonQBChau committed:
Update db_setup.py

SonQBChau committed:
Update schema.yml

SonQBChau committed:
Update schema.yml

SonQBChau committed:
Update schema.yml

SonQBChau committed:
Update schema.yml

SonQBChau committed:
Merge branch 'develop' into add_persons

SonQBChau committed:
remove drop FK

SonQBChau committed:
replace hardcode schema

SonQBChau committed:
Delete schema_ref.yml

SonQBChau committed:
Create db_test.sql

SonQBChau committed:
add limit 1 to optimize

SonQBChau committed:
remove duplication schema

SonQBChau committed:
use logger instead of print

OrdiNeu commented:
When I try to ingest a dataset with a person, there's no error message but:
```
(candig) 13:10 fnguyen@fnguyen-worktop:~/Documents/CanDIGv3$ curl -X PUT candig.docker.internal:5080/candig-api/v1/datasets -H "Content-Type: application/json" -H "Authorization: Bearer $TOKEN" -d @test-upload.json
(candig) 13:10 fnguyen@fnguyen-worktop:~/Documents/CanDIGv3$ docker logs candigv2_candig-api_1 --since=1m
2025-09-26 17:10:43,163 INFO sqlalchemy.engine.Engine BEGIN (implicit)
level: INFO, file: sqlalchemy.engine.Engine, log: BEGIN (implicit)
2025-09-26 17:10:43,163 INFO sqlalchemy.engine.Engine INSERT INTO candig.dataset (source_value, info) VALUES ($1::VARCHAR, $2::JSON) RETURNING candig.dataset.id
level: INFO, file: sqlalchemy.engine.Engine, log: INSERT INTO candig.dataset (source_value, info) VALUES ($1::VARCHAR, $2::JSON) RETURNING candig.dataset.id
2025-09-26 17:10:43,163 INFO sqlalchemy.engine.Engine [cached since 95.39s ago] ('asdf', '{"dataset_description": "test", "additionalProp1": {}}')
level: INFO, file: sqlalchemy.engine.Engine, log: [cached since 95.39s ago] ('asdf', '{"dataset_description": "test", "additionalProp1": {}}')
2025-09-26 17:10:43,165 INFO sqlalchemy.engine.Engine 
                        INSERT INTO omop.person (
                            gender_concept_id, year_of_birth, month_of_birth, 
                            day_of_birth, birth_datetime, race_concept_id, ethnicity_concept_id,
                            location_id, provider_id, care_site_id, person_source_value,
                            gender_source_value, gender_source_concept_id, race_source_value,
                            race_source_concept_id, ethnicity_source_value, ethnicity_source_concept_id
                        ) VALUES (
                            $1, $2, $3,
                            $4, $5, $6, $7,
                            $8, $9, $10, $11,
                            $12, $13, $14,
                            $15, $16, $17
                        ) RETURNING person_id
                    
level: INFO, file: sqlalchemy.engine.Engine, log: 
                        INSERT INTO omop.person (
                            gender_concept_id, year_of_birth, month_of_birth, 
                            day_of_birth, birth_datetime, race_concept_id, ethnicity_concept_id,
                            location_id, provider_id, care_site_id, person_source_value,
                            gender_source_value, gender_source_concept_id, race_source_value,
                            race_source_concept_id, ethnicity_source_value, ethnicity_source_concept_id
                        ) VALUES (
                            $1, $2, $3,
                            $4, $5, $6, $7,
                            $8, $9, $10, $11,
                            $12, $13, $14,
                            $15, $16, $17
                        ) RETURNING person_id
                    
2025-09-26 17:10:43,165 INFO sqlalchemy.engine.Engine [generated in 0.00016s] (0, 0, 0, 0, datetime.datetime(2025, 1, 1, 12, 10), 0, 0, 0, 0, 0, 'string', 'string', 0, 'string', 0, 'string', 0)
level: INFO, file: sqlalchemy.engine.Engine, log: [generated in 0.00016s] (0, 0, 0, 0, datetime.datetime(2025, 1, 1, 12, 10), 0, 0, 0, 0, 0, 'string', 'string', 0, 'string', 0, 'string', 0)
2025-09-26 17:10:43,169 INFO sqlalchemy.engine.Engine ROLLBACK
level: INFO, file: sqlalchemy.engine.Engine, log: ROLLBACK
level: ERROR, file: /app/candig-api/src/api/dataset_operations.py, log: {'message': 'Database Error in dataset.create: (sqlalchemy.dialects.postgresql.asyncpg.IntegrityError) <class \'asyncpg.exceptions.UniqueViolationError\'>: duplicate key value violates unique constraint "xpk_person"\nDETAIL:  Key (person_id)=(1) already exists.\n[SQL: \n                        INSERT INTO omop.person (\n                            gender_concept_id, year_of_birth, month_of_birth, \n                            day_of_birth, birth_datetime, race_concept_id, ethnicity_concept_id,\n                            location_id, provider_id, care_site_id, person_source_value,\n                            gender_source_value, gender_source_concept_id, race_source_value,\n                            race_source_concept_id, ethnicity_source_value, ethnicity_source_concept_id\n                        ) VALUES (\n                            $1, $2, $3,\n                            $4, $5, $6, $7,\n                            $8, $9, $10, $11,\n                            $12, $13, $14,\n                            $15, $16, $17\n                        ) RETURNING person_id\n                    ]\n[parameters: (0, 0, 0, 0, datetime.datetime(2025, 1, 1, 12, 10), 0, 0, 0, 0, 0, \'string\', \'string\', 0, \'string\', 0, \'string\', 0)]\n(Background on this error at: https://sqlalche.me/e/20/gkpj)'}
level: ERROR, file: connexion.middleware.exceptions, log: ProblemException(status_code=500, detail='An error occurred while creating the dataset in the database.')
(candig) 13:10 fnguyen@fnguyen-worktop:~/Documents/CanDIGv3$ curl candig.docker.internal:5080/candig-api/v1/datasets -H "Authorization: Bearer $TOKEN"
[]
```

Nobody seems to be imported

SonQBChau commented:
I couldn't recreate the bug. I'm using the develop branch for candigv3 and add_persons for candig-api. I did notice another bug where you need to manually delete the tmp/omop folder, otherwise the VOCAB won't be created properly.

Here's the JSON I used to insert:
```json
{
  "source_value": "unique_string",
  "persons": [
    {
      "gender_concept_id": 0,
      "year_of_birth": 0,
      "month_of_birth": 0,
      "day_of_birth": 0,
      "birth_datetime": "2000-01-01",
      "race_concept_id": 0,
      "ethnicity_concept_id": 0,
      "person_source_value": "string",
      "gender_source_value": "string",
      "gender_source_concept_id": 0,
      "race_source_value": "string",
      "race_source_concept_id": 0,
      "ethnicity_source_value": "string",
      "ethnicity_source_concept_id": 0
    }
  ],
  "info": {
    "dataset_description": "string",
    "additionalProp1": {}
  }
}
```

Can you clean the stack and test it again to check if the bug can be reproduced?

OrdiNeu commented:
Thank you for the `.json` -- it helped me diagnose exactly what was going on:
```
      "location_id": 0,
      "care_site_id": 0,
      "provider_id": 0,
```

With any of the above three entries, the database appears to crash out during upload with an error message:
```
{"type": "about:blank", "title": "Database Error", "detail": "An error occurred while creating the dataset in the database.", "status": 500}
```

and then in the logs you see the error I reported here: https://github.com/CanDIG/candig-api/pull/6#issuecomment-3339690359

Could we leave a more informative error message with the above error?

SonQBChau commented:
These are non-required foreign keys. You can leave them as null or provide a valid fk. In this case, 0 is not a valid value

OrdiNeu commented:
Neat. Could we leave a more informative error message with the above error?

SonQBChau commented:
I'm not sure I can give a better error message. I tried to capture all, but these are not specific and come from the framework

OrdiNeu commented:
I can't accept this unless we can make the error messages more reasonable -- otherwise we're going to end up with the same problems we had in CanDIGv2 where only Daisie and I were able to run DevOps, and all of our collaborators were confused at the error messages.

SonQBChau commented:
The input should pass through ETL and schema validation first. We haven't added those yet, but they should catch invalid input before it reaches the database.
I agree the current error message isn't helpful. What do you suggest approach?

OrdiNeu committed:
Add in more details into the response

OrdiNeu commented:
I added a commit `c69d0f1` to address it -- please take a look

SonQBChau commented:
I tested the PR, and I think I understand why you want that change. If you look at the API response, your fix now displays the error message directly in the response:
```json
{
  "type": "about:blank",
  "title": "Database Integrity Error",
  "detail": "Uploading the given dataset led to a violation of database integrity constraints.\nThe dataset id or person id in your input may not be unique.\nError details: (sqlalchemy.dialects.postgresql.asyncpg.IntegrityError) <class 'asyncpg.exceptions.UniqueViolationError'>: duplicate key value violates unique constraint \"dataset_source_value_key\"\nDETAIL:  Key (source_value)=(unique_string) already exists.\n[SQL: INSERT INTO candig.dataset (source_value, info) VALUES ($1::VARCHAR, $2::JSON) RETURNING candig.dataset.id]\n[parameters: ('unique_string', '{\"dataset_description\": \"string\", \"additionalProp1\": {}}')]\n(Background on this error at: https://sqlalche.me/e/20/gkpj)",
  "status": 500
}
```
However, if you look at the log, this adds nothing new to the developer as it already been logged and double logged: 
```json
level: ERROR, file: /Users/sonchau/Documents/DHDP/candig-api/src/api/dataset_operations.py, log: {'message': 'Database Integrity Error in dataset.put_by_id: (sqlalchemy.dialects.postgresql.asyncpg.IntegrityError) <class \'asyncpg.exceptions.UniqueViolationError\'>: duplicate key value violates unique constraint "dataset_source_value_key"\nDETAIL:  Key (source_value)=(unique_string) already exists.\n[SQL: INSERT INTO candig.dataset (source_value, info) VALUES ($1::VARCHAR, $2::JSON) RETURNING candig.dataset.id]\n[parameters: (\'unique_string\', \'{"dataset_description": "string", "additionalProp1": {}}\')]\n(Background on this error at: https://sqlalche.me/e/20/gkpj)'}
level: ERROR, file: connexion.middleware.exceptions, log: ProblemException(status_code=500, detail='Uploading the given dataset led to a violation of database integrity constraints.\nThe dataset id or person id in your input may not be unique.\nError details: (sqlalchemy.dialects.postgresql.asyncpg.IntegrityError) <class \'asyncpg.exceptions.UniqueViolationError\'>: duplicate key value violates unique constraint "dataset_source_value_key"\nDETAIL:  Key (source_value)=(unique_string) already exists.\n[SQL: INSERT INTO candig.dataset (source_value, info) VALUES ($1::VARCHAR, $2::JSON) RETURNING candig.dataset.id]\n[parameters: (\'unique_string\', \'{"dataset_description": "string", "additionalProp1": {}}\')]\n(Background on this error at: https://sqlalche.me/e/20/gkpj)')
```

This is more of a personal preference, but I don't think the API itself should expose detailed error to the end user. Normally, it should return a 500 status with maybe a short hint, but not the full details. Detailed logs are meant for developers, not end users, they don't need to see that information. While it's useful for debugging, I'm not sure it's a good practice in production. I haven't seen this kind of detailed error message in any public API.

OrdiNeu commented:
What if we remove the `{e}` error trace and just leave the hints + the name of the type of error, then?

SonQBChau commented:
We discussed this during the standup and agreed that it is ok to remove the SQL part from the error message.

SonQBChau committed:
Update dataset_operations.py

SonQBChau commented:
I pushed the update. I shortened the IntegrityError message and changed the status code to 400 to indicate bad user input. I also removed it from the delete operation since it wasn't needed there and stripped the detailed error info from the 500 response.

Now it shows:
```json
{
  "type": "about:blank",
  "title": "Database Integrity Error",
  "detail": "Violation of database integrity constraints.\nUnique value required.\nError details: Key (source_value)=(unique_string) already exists.",
  "status": 400
}
```
I think it clear enough and provides the user with the essential information. Let me know if you're okay with these changes.

OrdiNeu commented on src/api/dataset_operations.py:
```diff
@@ -153,21 +153,25 @@ async def create(body: dict):
+                extra_details += "Foreign key invalid.\n"
             if "uniqueviolationerror" in str(e).lower():
-                extra_details += "The dataset id or person id in your input may not be unique.\n"
+                extra_details += "Unique value required.\n"
```
"Value should be unique"

SonQBChau replied:
done

SonQBChau committed:
Update dataset_operations.py
